File: services/order-service-go/internal/events/envelope.go
package events

import (
	"fmt"
	"time"
)

// EventEnvelope represents the common envelope for all events.
// It is generic to allow strongly typed payloads per event.
type EventEnvelope[T any] struct {
	EventName     string    `json:"eventName"`
	EventVersion  int       `json:"eventVersion"`
	EventID       string    `json:"eventId"`
	CorrelationID string    `json:"correlationId,omitempty"`
	CausationID   string    `json:"causationId,omitempty"`
	Producer      string    `json:"producer"`
	PartitionKey  string    `json:"partitionKey"`
	Sequence      *int64    `json:"sequence,omitempty"`
	OccurredAt    time.Time `json:"occurredAt"`
	Schema        string    `json:"schema"`
	Payload       T         `json:"payload"`
}

// EnvelopeMetadata carries correlation/causation context for emitted events.
type EnvelopeMetadata struct {
	CorrelationID string
	CausationID   string
}

// Validate ensures the envelope contains the expected event identity.
func (e EventEnvelope[T]) Validate(expectedName string, expectedVersion int) error {
	if e.EventName != expectedName {
		return fmt.Errorf("unexpected eventName: %s", e.EventName)
	}
	if e.EventVersion != expectedVersion {
		return fmt.Errorf("unexpected eventVersion: %d", e.EventVersion)
	}
	if e.PartitionKey == "" {
		return fmt.Errorf("missing partitionKey")
	}
	return nil
}

File: services/payment-service-dotnet/src/PaymentService/Contracts/Envelope.cs
namespace PaymentService.Contracts
{
    public sealed class Envelope<TPayload>
    {
        public required string EventName { get; init; }
        public required int EventVersion { get; init; }
        public required Guid EventId { get; init; }
        public Guid? CorrelationId { get; init; }
        public required string Producer { get; init; }
        public string? PartitionKey { get; init; }
        public long? Sequence { get; init; }
        public required DateTimeOffset OccurredAt { get; init; }
        public string? Schema { get; init; }
        public required TPayload Payload { get; init; }
    }

}